Update private repository detection logic for Gitlab and Github

erzhtor opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/27382

#### Description

This PR: 
- Updates private repository detection logic for Github & Gitlab by using corresponding REST APIs

> NOTE: it still sends repo names to check if the repo exists in Sourcegraph or not.

#### Before merging

- [ ] Test on different code hosts (if applicable)
    - [ ] GitHub
    - [ ] Gitlab
    - [ ] GitHub Enterprise
    - [ ] Refined GitHub
    - [ ] Phabricator
    - [ ] Phabricator integration
    - [ ] Bitbucket
    - [ ] Bitbucket integration

- [ ] Test on different browsers (if applicable)
    - [ ] Chrome
    - [ ] Firefox
    - [ ] Safari
- [ ] Add change log message to [client/browser/CHANGELOG.md](./client/browser/CHANGELOG.md), under "Unreleased" section. (if applicable)

muratsu commented on client/shared/src/util/fetchCache.ts:
```diff
@@ -0,0 +1,60 @@
+const cache = new Map<string, FetchCacheReturnType<any>>()
+const runningRequests = new Map<string, Promise<FetchCacheReturnType<any>>>()
+const INVALIDATE_TIMEOUT = 60 * 1000 // 1 minute
```
I think we can cache this value for even longer. The GitHub API limit can be fairly limiting:

> For unauthenticated requests, the rate limit allows for up to 60 requests per hour. Unauthenticated requests are associated with the originating IP address, and not the user making requests.

muratsu replied:
Thoughts @tjkandala?

tjkandala replied:
Yeah I agree, it's not likely that a repo's visibility would change during this time either. We can parameterize `INVALIDATE_TIMEOUT` (using the value from the first call for that key when cached) and make the default value a lot longer (since this is the only place we use `fetchCache` at the moment)

erzhtor replied:
Current fetchCache does it in memory, meaning on page reload, it will lose its cache, so I guess we need to make it persistent or do it in the background or save it to localStorage.

However, even with persistent caching, I'm just thinking about the edge case when the user might visit more than 60 repositories within an hour 🤷 ?

tjkandala replied:
Would it be possible to use personal access tokens to work around that?

erzhtor replied:
I think it won't help since the token has a rate of 5k which will be shared across all browser extension users.
https://docs.github.com/en/developers/apps/building-github-apps/rate-limits-for-github-apps#user-to-server-requests

erzhtor replied:
FYI, added DOM check fallback for GH in case of rate limit.

muratsu commented on client/browser/src/shared/code-hosts/github/codeHost.ts:
```diff
@@ -367,31 +368,32 @@ const searchEnhancement: CodeHost['searchEnhancement'] = {
+    return fetchCache<{ private?: boolean }>(`https://api.github.com/repos/${repoName}`)
+        .then(({ data }) => typeof data.private !== 'boolean' || data.private)
+        .catch(error => {
+            console.warn('Failed to fetch if the repository is private.', error)
```
It would be nice to log this error to sentry if it's available. We want to see if we're hitting the rate-limiting frequently.

muratsu commented on client/browser/src/shared/code-hosts/gitlab/codeHost.ts:
```diff
@@ -129,17 +130,34 @@ const notificationClassNames = {
+    )
+        .then(({ data }) => data?.visibility !== 'public')
+        .catch(error => {
+            console.warn('Failed to fetch if the repository visibility.', error)
```
It would be nice to log this error to sentry if it's available.

tjkandala commented on client/browser/src/shared/code-hosts/shared/codeHost.tsx:
```diff
@@ -165,7 +165,7 @@ export interface CodeHost extends ApplyLinkPreviewOptions {
      * Basic contextual information for the current code host.
      */
-    getContext?: () => CodeHostContext
+    getContext?: () => Promise<CodeHostContext>
```
How do you feel about allowing either `Promise<CodeHostContext>` or just `CodeHostContext` and wrapping the return value with `Promise.resolve` as needed so we don't have to do that for code hosts in which we can synchronously resolve context? (ignore if we plan on querying APIs for all the code hosts eventually)

erzhtor replied:
I think it is done so in the current implementation. All other code hosts return Promise.resolve.

tjkandala commented on client/shared/src/util/fetchCache.test.ts:
```diff
@@ -0,0 +1,43 @@
+import fetch from 'jest-fetch-mock'
```
🔥️

tjkandala commented on client/shared/src/util/fetchCache.ts:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    const key = JSON.stringify(args)
+    if (cache.has(key)) {
+        return cache.get(key) as FetchCacheReturnType<T>
+    }
+
+    if (runningRequests.has(key)) {
+        return (await runningRequests.get(key)) as FetchCacheReturnType<T>
+    }
```
Since the return value will end up being a `Promise` anyways, could we consolidate these maps?

erzhtor replied:
Thanks, it seems to work! 👍

tjkandala commented on client/shared/src/util/fetchCache.ts:
```diff
@@ -0,0 +1,60 @@
+            setTimeout(() => cache.delete(key), INVALIDATE_TIMEOUT)
+
+            return result
+        })
+        .finally(() => runningRequests.delete(key))
```
If we end up consolidating the maps, we can instead delete `Promise`s from the map in `catch` (in case of network error)

tjkandala commented on client/browser/src/shared/code-hosts/gitlab/codeHost.ts:
```diff
@@ -129,17 +130,34 @@ const notificationClassNames = {
+    if (window.location.hostname !== 'gitlab.com' || !projectId) {
+        return Promise.resolve(true)
+    }
+    return fetchCache<{ visibility?: 'public' | 'private' | 'internal' }>(
+        `https://gitlab.com/api/v4/projects/${projectId}`
```
Can you add a link to the documentation for this endpoint (if it exists)?

tjkandala commented on client/browser/src/shared/code-hosts/github/codeHost.ts:
```diff
@@ -367,31 +368,32 @@ const searchEnhancement: CodeHost['searchEnhancement'] = {
+    if (window.location.hostname !== 'github.com') {
+        return Promise.resolve(true)
+    }
+    return fetchCache<{ private?: boolean }>(`https://api.github.com/repos/${repoName}`)
+        .then(({ data }) => typeof data.private !== 'boolean' || data.private)
```
Same GitLab comment re documentation

tjkandala reviewed (approved):
Great work!! Nice solution to a longstanding weak spot in our browser extension

erzhtor commented:
@sourcegraph/extensibility, thanks for the reviews. I've updated PR:
- Address PR comments
- Moved fetchCache to background script for persistence caching
- Added DOM check fallback for GitHub in case of rate limit exceeded

I would appreciate your feedback/re-reviews.

muratsu commented on client/browser/src/shared/code-hosts/github/codeHost.ts:
```diff
@@ -367,31 +369,55 @@ const searchEnhancement: CodeHost['searchEnhancement'] = {
+            const rateLimit = response.headers['x-ratelimit-remaining']
+            if (Number(rateLimit) <= 0) {
+                const error = new Error('Github rate limit exceeded.')
+                Sentry.captureException(error)
```
Just double confirming that this only works if user [allows error reporting](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/browser/src/shared/sentry/index.ts?L39)

erzhtor replied:
Yep, AFAIU, Sentry disabling works globally, so any captured exceptions won't be sent to sentry if it is disabled.

muratsu commented:
Left one comment LGTM otherwise
